keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -219,76 +210,31 @@ class SqlArtifactRepository(
-        .from(ARTIFACT_VERSIONS)
-        .where(ARTIFACT_VERSIONS.NAME.eq(name))
-        .and(ARTIFACT_VERSIONS.TYPE.eq(type))
-        .fetch()
-        .getValues(ARTIFACT_VERSIONS.VERSION)
-    }
-  }
-
-  override fun versions(artifact: DeliveryArtifact, limit: Int): List<String> {
+  override fun versions(artifact: DeliveryArtifact, limit: Int): List<PublishedArtifact> {
     if (!isRegistered(artifact.name, artifact.type)) {
       throw NoSuchArtifactException(artifact)
     }
 
-    val versions = sqlRetry.withRetry(READ) {
+    return sqlRetry.withRetry(READ) {
       jooq
-        .select(ARTIFACT_VERSIONS.VERSION, ARTIFACT_VERSIONS.RELEASE_STATUS)
+        .select(
+          ARTIFACT_VERSIONS.NAME,
+          ARTIFACT_VERSIONS.TYPE,
+          ARTIFACT_VERSIONS.VERSION,
+          ARTIFACT_VERSIONS.RELEASE_STATUS,
+          ARTIFACT_VERSIONS.CREATED_AT,
+          ARTIFACT_VERSIONS.GIT_METADATA,
+          ARTIFACT_VERSIONS.BUILD_METADATA
+        )
         .from(ARTIFACT_VERSIONS)
         .where(ARTIFACT_VERSIONS.NAME.eq(artifact.name))
         .and(ARTIFACT_VERSIONS.TYPE.eq(artifact.type))
-        .apply {
-          if (artifact.filteredByReleaseStatus) {
-            and(ARTIFACT_VERSIONS.RELEASE_STATUS.`in`(*artifact.statuses.map { it.toString() }.toTypedArray()))
-          } else {
-            // TODO: should we also be comparing the repo with what's configured for the app in front50?
-
-            if (artifact.filteredByPullRequest) {
-              and(ARTIFACT_VERSIONS_PR_NUMBER.isNotNull).and(ARTIFACT_VERSIONS_PR_NUMBER.ne(EMPTY_PR_NUMBER))
-            }
-
-            if (artifact.filteredByBranch) {
-              artifact.from?.branch?.name?.also {
-                and(ARTIFACT_VERSIONS_BRANCH.eq(it))
-              }
-              artifact.from?.branch?.startsWith?.also {
-                and(ARTIFACT_VERSIONS_BRANCH.startsWith(it))
-              }
-              artifact.from?.branch?.regex?.also {
-                and(ARTIFACT_VERSIONS_BRANCH.likeRegex(it))
-              }
-            }
-
-            // With branches or pull requests, delegate sorting and limiting to the database
-            if (artifact.filteredByPullRequest || artifact.filteredByBranch) {
-              and(ARTIFACT_VERSIONS.CREATED_AT.isNotNull)
-                .orderBy(ARTIFACT_VERSIONS.CREATED_AT.desc())
-                .limit(limit)
-            }
-          }
-        }
-        .fetch()
-        .getValues(ARTIFACT_VERSIONS.VERSION)
-    }
-
-    return if (artifact.filteredByPullRequest || artifact.filteredByBranch) {
-      versions
-    } else {
-      val sortedVersions = versions.sortedWith(artifact.versioningStrategy.comparator)
-      if (artifact is DockerArtifact) {
-        filterDockerVersions(artifact, sortedVersions, limit)
-      } else {
-        sortedVersions.subList(0, Math.min(sortedVersions.size, limit))
-      }
+        .fetchSortedArtifactVersions(artifact, limit)
```

Comment:
This applies native SQL sorting and limiting when applicable, and falls back to the comparator from the sorting strategy when that doesn't apply. This means the return from this function is pre-sorted.